stubs/pyjks/jks/util.pyi

```diff
@@ -0,0 +1,66 @@
+from collections.abc import Iterable, Mapping
+from pathlib import Path
+from struct import Struct
+from typing import Any
+from typing_extensions import Final, Literal, Self, TypeAlias
+
+from .bks import BksKeyEntry
+from .jks import PrivateKeyEntry
+
+b8: Final[Struct]
+b4: Final[Struct]
+b2: Final[Struct]
+b1: Final[Struct]
+py23basestring: Final[tuple[type[str], type[str]]]
+RSA_ENCRYPTION_OID: Final[tuple[int, ...]]
+DSA_OID: Final[tuple[int, ...]]
+DSA_WITH_SHA1_OID: Final[tuple[int, ...]]
+
+_KeystoreType: TypeAlias = Literal["jks", "jceks", "bks", "uber"]
+_PemType: TypeAlias = Literal["CERTIFICATE", "PUBLIC KEY", "PRIVATE KEY", "RSA PRIVATE KEY"]
+
+class KeystoreException(Exception): ...
+class KeystoreSignatureException(KeystoreException): ...
+class DuplicateAliasException(KeystoreException): ...
+class NotYetDecryptedException(KeystoreException): ...
+class BadKeystoreFormatException(KeystoreException): ...
+class BadDataLengthException(KeystoreException): ...
+class BadPaddingException(KeystoreException): ...
+class BadHashCheckException(KeystoreException): ...
+class DecryptionFailureException(KeystoreException): ...
+class UnsupportedKeystoreVersionException(KeystoreException): ...
+class UnexpectedJavaTypeException(KeystoreException): ...
+class UnexpectedAlgorithmException(KeystoreException): ...
+class UnexpectedKeyEncodingException(KeystoreException): ...
+class UnsupportedKeystoreTypeException(KeystoreException): ...
+class UnsupportedKeystoreEntryTypeException(KeystoreException): ...
+class UnsupportedKeyFormatException(KeystoreException): ...
+
+class AbstractKeystore:
+    store_type: _KeystoreType
+    entries: Mapping[str, AbstractKeystoreEntry]
+    def __init__(self, store_type: _KeystoreType, entries: dict[str, AbstractKeystoreEntry]) -> None: ...
```

Comment:
I think this might be back to front: it looks to me like you can provide any mapping to `__init__`, but it will always be converted to a `dict` before being set as an instance attribute: https://github.com/kurtbrose/pyjks/blob/0a046e54337a0c271ef5f7fdd9aa79c5ad350486/jks/util.py#L78-L80

```suggestion
    entries: dict[str, AbstractKeystoreEntry]
    def __init__(self, store_type: _KeystoreType, entries: Mapping[str, AbstractKeystoreEntry]) -> None: ...
```

You could probably be even more permissive and use `_typeshed.SupportsKeysAndGetItem`, since that's all that `builtins.dict` needs for its constructor:

https://github.com/python/typeshed/blob/16b98846017c07ebbca9c75c099cb4b6c6b11ad8/stdlib/builtins.pyi#L1063-L1071

```suggestion
    entries: dict[str, AbstractKeystoreEntry]
    def __init__(self, store_type: _KeystoreType, entries: SupportsKeysAndGetItem[str, AbstractKeystoreEntry]) -> None: ...
```

Reply:
The reason for hinting `AbstractKeystore.entries: Mapping` is because a the subclass `KeyStore` overrides `entries` and mypy complained about mutability.

But I'm fine with using `type: ignore[override]` to suppress that.

Reply:
Aha, got it. I'm that case, I'm fine with what using `Mapping` for the `entries` attribute (but please add an explanatory comment). I'd also be happy with changing it to `dict` and adding the `type: ignore`; will defer to you on which one you'd prefer!

Reply:
I don't have strong feelings, but using `dict` seems more consistent.